odiglet/pkg/kube/runtime_details/instrumentationconfigs_controller.go

```diff
@@ -0,0 +1,136 @@
+package runtime_details
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	odigosv1 "github.com/odigos-io/odigos/api/odigos/v1alpha1"
+	k8sutils "github.com/odigos-io/odigos/k8sutils/pkg/utils"
+	kubeutils "github.com/odigos-io/odigos/odiglet/pkg/kube/utils"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type instrumentationConfigPredicate struct{}
+
+func (p *instrumentationConfigPredicate) Create(e event.CreateEvent) bool {
+	obj, ok := e.Object.(*odigosv1.InstrumentationConfig)
+	if !ok {
+		return false
+	}
+	// we only care about new InstrumentationConfig objects
+	// the event will be triggered also when odiglet starts, and the instrumentation config is created
+	// in controller-runtime cache
+	// checking the RuntimeDetailsByContainer map should filter cases where we do not want to trigger the reconcile
+	// for workload which has already been reconciled previously
+	return len(obj.Status.RuntimeDetailsByContainer) == 0
+}
+
+func (p *instrumentationConfigPredicate) Update(e event.UpdateEvent) bool {
+	return false
+}
+
+func (p *instrumentationConfigPredicate) Delete(e event.DeleteEvent) bool {
+	return false
+}
+
+func (p *instrumentationConfigPredicate) Generic(e event.GenericEvent) bool {
+	return false
+}
+
+type InstrumentationConfigReconciler struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	Clientset *kubernetes.Clientset
+}
+
+func (r *InstrumentationConfigReconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+
+	logger := log.FromContext(ctx)
+
+	var instrumentationConfig odigosv1.InstrumentationConfig
+	err := r.Get(ctx, request.NamespacedName, &instrumentationConfig)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if len(instrumentationConfig.OwnerReferences) != 1 {
+		return reconcile.Result{}, fmt.Errorf("InstrumentationConfig %s/%s has %d owner references, expected 1", instrumentationConfig.Namespace, instrumentationConfig.Name, len(instrumentationConfig.OwnerReferences))
+	}
+
+	workload, labels, err := getWorkloadAndLabelsfromOwner(ctx, r.Client, instrumentationConfig.Namespace, instrumentationConfig.OwnerReferences[0])
+	if err != nil {
+		logger.Error(err, "Failed to get workload and labels from owner")
+		return reconcile.Result{}, err
+	}
+
+	pods, err := kubeutils.GetRunningPods(ctx, labels, workload.GetNamespace(), r.Client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// we need to apply runtime detection for just one pod with newer generation
+	var podForRuntimeDetails *corev1.Pod
+	selectedPodGeneration := int64(0)
+	for _, pod := range pods {
+		podGeneration, err := GetPodGeneration(ctx, r.Clientset, &pod)
```

Comment:
Here we are using the "slow" client. We can have these loop iterations done in parallel and save time. I'm not sure if this will become very slow in large clusters or not. Assuming 10ms per API server query and 100 pods will result in 1 second for this loop.

Reply:
It's only for pods of this workload, that run on the current node.
I think this will most times be 0, or 1-2, and in rare cases it might be more, but not something we should be worried about.